configure.ac

```diff
@@ -1635,6 +1635,10 @@ SQUID_CHECK_LIB_WORKS(ldap,[
     AC_CHECK_HEADERS(mozldap/ldap.h)
     SQUID_CHECK_LDAP_API
   ])
+  # We can still rely on deprecated APIs on MacOS
+  AS_IF([test "x$squid_host_os.$ac_cv_header_ldap_h.$ac_cv_lib_ldap_ldap_init" = "xdarwin.yes.yes"],[
+    SQUID_CC_ADD_CXXFLAG_WARNING_IF_SUPPORTED([-Wno-deprecated-declarations])
```

Comment:
This call adjusts CXXFLAGS to enable warning-free compilation of code that uses "native" MacOS LDAP headers. This call sits inside SQUID_CHECK_LIB_WORKS(ldap...). I believe this call location is wrong (even if it "works"):

* If adjusted flags are needed for SQUID_CHECK_LDAP_API (or other "Does ldap work?" checks above SQUID_CHECK_LDAP_API) to work correctly, then the flags should be adjusted **earlier**, before executing the code that needs them.
* Otherwise, the adjusted flags are only needed for Squid code that uses LDAP APIs. In that case, they should be adjusted **later**, after SQUID_CHECK_LIB_WORKS() concludes that "ldap works" (i.e. sets LIBLDAP_LIBS) and exits. If this bullet is applicable, the (moved) AS_IF condition will need to test LIBLDAP_LIBS instead of ac_cv_....

N.B. In the above "X needs flags" stands for "X _may_ fail to compile without -Wno-deprecated-declarations". I do understand that we are not checking whether X _will necessarily_ fail without that flag, and I am not requesting that we start checking that.

Reply:
> This call adjusts SQUID_CXXFLAGS to enable warning-free compilation of code that uses "native" MacOS LDAP headers. This call sits inside SQUID_CHECK_LIB_WORKS(ldap...). I believe this call location is wrong (even if it "works"):

You are right, but not for the reasons you think.
Configure tests pass without the flag. Build doesn't. The reason is that configure doesn't set CXXFLAGS to -Werror; it sets it in SQUID_CXXFLAGS which are not used for the tests but are used for the build. I don't dare touch that, it's likely to be a big can of worms; I've opened draft PR #1702 to facilitate the discussion about this

> * If adjusted flags are needed for SQUID_CHECK_LDAP_API (or other "Does ldap work?" checks above SQUID_CHECK_LDAP_API) to work correctly, then the flags should be adjusted **earlier**, before executing the code that needs them.

After PR 1702, lands (if ever), yes. Right now doing that would result in enabling this flag unconditionally, unless the user were to specify --without-{ldap,sasl} (there are workarounds for this, but it'd be adding more and more complication)

> * Otherwise, the adjusted flags are only needed for Squid code that uses LDAP APIs. In that case, they should be adjusted **later**, after SQUID_CHECK_LIB_WORKS() concludes that "ldap works" (i.e. sets LIBLDAP_LIBS) and exits. If this bullet is applicable, the (moved) AS_IF condition will need to test LIBLDAP_LIBS instead of ac_cv_....

Yes, done this.

> N.B. In the above "X needs flags" stands for "X _may_ fail to compile without -Wno-deprecated-declarations". I do understand that we are not checking whether X _will necessarily_ fail without that flag, and I am not requesting that we start checking that.

Yes